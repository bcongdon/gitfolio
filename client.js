import { ApolloClient } from "apollo-client";
import {
  InMemoryCache,
  IntrospectionFragmentMatcher
} from "apollo-cache-inmemory";
import gql from "graphql-tag";
import { createHttpLink } from "apollo-link-http";
import { setContext } from "apollo-link-context";
import schema from "./github_graphql_schema";
import axios from "axios";
import { AUTH_TOKEN } from "./secrets";

/**
 * authLink sets up the Authentication necessary to make requests to the Github GraphQL API
 */
const authLink = setContext((_, { headers }) => {
  return {
    headers: {
      ...headers,
      authorization: AUTH_TOKEN ? `Bearer ${AUTH_TOKEN}` : ""
    }
  };
});

/**
 * The query for pulling a complete Github profile (user data, repos, followers / following)
 */
const PullGithubProfileQuery = gql`
  query PullGithubProfileQuery($username: String!) {
    user(login: $username) {
      name
      login
      bio
      createdAt
      location
      email
      websiteUrl
      viewerCanFollow
      viewerIsFollowing
      avatarUrl
      repositories(
        first: 100
        privacy: PUBLIC
        affiliations: OWNER
        orderBy: { direction: DESC, field: PUSHED_AT }
      ) {
        nodes {
          name
          description
          viewerHasStarred
          primaryLanguage {
            name
          }
          stargazers {
            totalCount
          }
          forkCount
          pushedAt
          owner {
            login
          }
          url
        }
        totalCount
      }
      followers(first: 100) {
        nodes {
          login
          name
          avatarUrl
          url
          viewerIsFollowing
        }
        totalCount
      }
      following(first: 100) {
        nodes {
          login
          name
          avatarUrl
          url
          viewerIsFollowing
        }
        totalCount
      }
    }
  }
`;

const SearchRepositories = gql`
  query SearchRepositories($query: String!) {
    search(query: $query, type: REPOSITORY, first: 25) {
      nodes {
        ... on Repository {
          name
          description
          viewerHasStarred
          primaryLanguage {
            name
          }
          stargazers {
            totalCount
          }
          forkCount
          description
          url
          pushedAt
          owner {
            login
          }
        }
      }
    }
  }
`;

const SearchUsers = gql`
  query SearchUsers($query: String!) {
    search(query: $query, type: USER, first: 25) {
      nodes {
        ... on User {
          name
          login
          bio
          createdAt
          location
          email
          websiteUrl
          viewerCanFollow
          viewerIsFollowing
          avatarUrl
        }
      }
    }
  }
`;

const fragmentMatcher = new IntrospectionFragmentMatcher({
  introspectionQueryResultData: schema
});

const apolloClient = new ApolloClient({
  link: authLink.concat(
    createHttpLink({
      uri: "https://api.github.com/graphql"
    })
  ),
  cache: new InMemoryCache({ fragmentMatcher })
});

/**
 * Pulls a Github Profile from the Github GraphQL API
 * @param  {string} username The username of the github user to pull
 * @return {Promise}         A promise with the pulled data
 */
function getProfile(username) {
  return apolloClient.query({
    query: PullGithubProfileQuery,
    variables: { username },
    fetchPolicy: "network-only"
  });
}

function searchUsers(query) {
  return apolloClient.query({
    query: SearchUsers,
    variables: { query },
    fetchPolicy: "network-only"
  });
}

function searchRepos(query) {
  return apolloClient.query({
    query: SearchRepositories,
    variables: { query },
    fetchPolicy: "network-only"
  });
}

/**
 * Sets whether or not the current user is starting the provided repo
 * @param {string} owner    The repo owner's login
 * @param {string} repo     The repo name
 * @param {bool}   newValue Whether or not the current user should be starring the given repo
 */
function setStarringRepository(owner, repo, newValue) {
  let method = newValue ? "PUT" : "DELETE";
  return axios({
    method,
    url: `https://api.github.com/user/starred/${owner}/${repo}`,
    params: {
      access_token: AUTH_TOKEN
    }
  });
}

/**
 * Sets whether or not the current user is following the provided user
 * @param {string} user     The user to follow/unfollow
 * @param {string} newValue Whether or not the current user should be following the provided user
 */
function setFollowingUser(user, newValue) {
  let method = newValue ? "PUT" : "DELETE";
  return axios({
    method,
    url: `https://api.github.com/user/following/${user}`,
    params: {
      access_token: AUTH_TOKEN
    }
  });
}

function getNotifications() {
  return axios({
    url: `https://api.github.com/notifications`,
    params: { access_token: AUTH_TOKEN }
  });
}

function getCommitActivity(owner, repo) {
  return axios({
    url: `https://api.github.com/repos/${
      owner.login
    }/${repo}/stats/commit_activity`,
    params: { access_token: AUTH_TOKEN }
  });
}

export default {
  getProfile,
  setStarringRepository,
  setFollowingUser,
  getNotifications,
  searchRepos,
  searchUsers,
  getCommitActivity
};
