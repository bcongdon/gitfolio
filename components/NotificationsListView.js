import React from "react";
import client from "../client";
import { Linking } from "react-native";
import {
  Container,
  Content,
  Text,
  Header,
  Body,
  Title,
  List,
  ListItem,
  Spinner,
  Icon,
  Left,
  Right
} from "native-base";
import moment from "moment";

/**
 * Icon names for notification subjects
 */
const SUBJECT_ICONS = {
  Issue: "issue-opened",
  Commit: "git-commit",
  PullRequest: "git-pull-request"
};

/**
 * Icon colors by notification subject
 */
const ICON_COLORS = {
  Issue: "#00a74c",
  PullRequest: "#00a74c"
};

export default class NotificationsListView extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      notifications: [],
      loading: true
    };
  }

  componentWillMount() {
    client.getNotifications().then(data => {
      this.setState({
        notifications: data.data,
        loading: false
      });
    });
  }

  /**
   * Renders an individual notification as a ListItem
   * @param  {object}  notification The object representing the notification
   * @return {ListItem}             The rendered notification
   */
  renderNotification(notification) {
    let iconName = SUBJECT_ICONS[notification.subject.type] || "bell";
    let iconColor = ICON_COLORS[notification.subject.type] || "#c8c9cb";
    return (
      <ListItem
        button
        key={notification.id}
        onPress={() => Linking.openURL(notification.repository.html_url)}
      >
        <Left>
          <Icon
            style={{ color: iconColor, fontSize: 20 }}
            type="Octicons"
            name={iconName}
            ios={iconName}
          />
          <Text style={{ paddingLeft: 10 }}>
            {notification.subject.title}
            {"\n"}
            <Text note>{notification.repository.full_name}</Text>
          </Text>
        </Left>
        <Right style={{ flex: 0.5 }}>
          <Text note>{moment(notification.updated_at).fromNow()}</Text>
        </Right>
      </ListItem>
    );
  }

  render() {
    var notifications;

    // Display a loading spinner if the component is in the loading state
    if (this.state.loading) {
      notifications = <Spinner />;
    } else if (this.state.notifications.length > 0) {
      // Display the list of notifications if the component has notifications to show
      notifications = (
        <List>{this.state.notifications.map(this.renderNotification)}</List>
      );
    } else {
      // Display a message if there are no notifications
      notifications = <Text style={{paddingTop: 10, textAlign: 'center' }} note>No new notifications</Text>;
    }

    return (
      <Container>
        <Header>
          <Body>
            <Title>Notifications</Title>
          </Body>
        </Header>
        <Content>{notifications}</Content>
      </Container>
    );
  }
}
