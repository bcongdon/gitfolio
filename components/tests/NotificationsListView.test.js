import NotificationsListView from "../NotificationsListView";
import React from "react";
import { Spinner, ListItem } from "native-base";
import client from "../../client";

import renderer from "react-test-renderer";

it("renders without crashing", () => {
  let component = <NotificationsListView />;

  const rendered = renderer.create(component).toJSON();
  expect(rendered).toBeTruthy();
});

it("calls API on component mount", () => {
  client.getNotifications = jest
    .genMockFn()
    .mockReturnValue(new Promise(() => {}));

  let component = <NotificationsListView />;
  renderer.create(component);

  expect(client.getNotifications).toBeCalled();
  expect(client.getNotifications.mock.calls.length).toEqual(1);
});

it("should pass snapshot test with notifications", () => {
  let component = <NotificationsListView />;

  let dummyNotification = {
    subject: {
      type: "Issue",
      title: "Issue title"
    },
    repository: { full_name: "some_repo" },
    id: 123
  };

  const renderTree = renderer.create(component);
  const rendered = renderTree.root;
  rendered.instance.setState({
    notifications: [dummyNotification],
    loading: false
  });

  jest.runAllTicks();
  const list = rendered.findByType(ListItem);

  expect(list).toBeTruthy();
  expect(renderTree.toJSON()).toMatchSnapshot();
});

it("should display a loading spinner on mount", () => {
  let component = <NotificationsListView />;
  const renderTree = renderer.create(component);
  let rendered = renderTree.root;

  expect(rendered.instance.state.loading).toBeTruthy();
  expect(rendered.findByType(Spinner)).toBeTruthy();
});
