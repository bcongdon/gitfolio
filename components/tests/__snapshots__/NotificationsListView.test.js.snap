// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should pass snapshot test with notifications 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "height": 1334,
    }
  }
  virtual={undefined}
>
  <View>
    <View
      style={
        Object {
          "backgroundColor": "#F8F8F8",
          "borderBottomColor": "#a7a6ab",
          "borderBottomWidth": 0.5,
          "elevation": 3,
          "flexDirection": "row",
          "height": 64,
          "justifyContent": "center",
          "left": 0,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 15,
          "right": 0,
          "shadowColor": undefined,
          "shadowOffset": undefined,
          "shadowOpacity": undefined,
          "shadowRadius": undefined,
          "top": 0,
        }
      }
      virtual={undefined}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "center",
            "flex": 1,
          }
        }
        virtual={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 17,
              "fontWeight": "600",
              "textAlign": "center",
            }
          }
          virtual={undefined}
        >
          Notifications
        </Text>
      </View>
    </View>
  </View>
  <RCTScrollView
    automaticallyAdjustContentInsets={false}
    contentContainerStyle={undefined}
    contentInset={
      Object {
        "bottom": 0,
      }
    }
    enableAutoAutomaticScroll={true}
    enableResetScrollToCoords={true}
    extraHeight={75}
    extraScrollHeight={0}
    getScrollResponder={[Function]}
    handleOnScroll={[Function]}
    keyboardDismissMode="interactive"
    keyboardOpeningTime={250}
    keyboardShouldPersistTaps="handled"
    keyboardSpace={0}
    onScroll={[Function]}
    resetKeyboardSpace={[Function]}
    resetScrollToCoords={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    scrollEventThrottle={1}
    scrollForExtraHeightOnAndroid={[Function]}
    scrollToEnd={[Function]}
    scrollToFocusedInput={[Function]}
    scrollToPosition={[Function]}
    showsVerticalScrollIndicator={true}
    style={
      Object {
        "backgroundColor": "transparent",
        "flex": 1,
      }
    }
    viewIsInsideTabBar={false}
    virtual={undefined}
  >
    <View>
      <View
        closeOnRowBeginSwipe={false}
        closeOnRowPress={true}
        closeOnScroll={true}
        directionalDistanceChangeThreshold={2}
        disableLeftSwipe={false}
        disableRightSwipe={false}
        leftOpenValue={0}
        previewFirstRow={false}
        recalculateHiddenLayout={false}
        rightOpenValue={0}
        style={Object {}}
        swipeToOpenPercent={50}
        virtual={undefined}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={null}
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View
            button={true}
            onPress={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderBottomWidth": 0.5,
                "borderColor": "#c9c9c9",
                "flexDirection": "row",
                "marginLeft": 16,
                "paddingBottom": 13,
                "paddingRight": 16,
                "paddingTop": 13,
              }
            }
            virtual={undefined}
          >
            <View
              style={
                Object {
                  "alignItems": "flex-start",
                  "alignSelf": "center",
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
              virtual={undefined}
            >
              <Text
                accessible={true}
                allowFontScaling={false}
                ellipsizeMode="tail"
                ios="issue-opened"
                style={
                  Array [
                    Object {
                      "color": undefined,
                      "fontSize": 12,
                    },
                    Array [
                      Object {
                        "color": "#000",
                        "fontSize": 30,
                      },
                      Object {
                        "color": "#000",
                        "fontSize": 20,
                        "width": 20,
                      },
                      Object {
                        "color": "#00a74c",
                        "fontSize": 20,
                      },
                    ],
                    Object {
                      "fontFamily": "octicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                  ]
                }
                type="Octicons"
                virtual={undefined}
              >
                ï…”
              </Text>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "alignSelf": "center",
                      "color": "#000",
                      "fontFamily": "System",
                      "fontSize": 16,
                    },
                    Object {
                      "paddingLeft": 10,
                    },
                  ]
                }
                uppercase={false}
                virtual={undefined}
              >
                Issue title
                

                <Text
                  accessible={true}
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                  note={true}
                  style={
                    Object {
                      "color": "#a7a7a7",
                      "fontFamily": "System",
                      "fontSize": 14,
                    }
                  }
                  uppercase={false}
                  virtual={undefined}
                >
                  some_repo
                </Text>
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "flex-end",
                    "alignSelf": "center",
                    "flex": 0.28,
                    "paddingBottom": null,
                    "paddingLeft": null,
                    "paddingRight": null,
                    "paddingTop": null,
                  },
                  Object {
                    "flex": 0.5,
                  },
                ]
              }
              virtual={undefined}
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                note={true}
                style={
                  Object {
                    "alignSelf": null,
                    "color": "#808080",
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "200",
                  }
                }
                uppercase={false}
                virtual={undefined}
              >
                a few seconds ago
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
